src/main/scala/chisel3/UnitTestMain.scala

```diff
@@ -0,0 +1,109 @@
+    var shouldExit = false
+    val parser = new OptionParser[Config]("chisel3.UnitTests") {
+      head("Chisel Unit Test Utility")
+      help("help").abbr("h")
+
+      opt[File]('o', "output")
+        .text("Output file name (\"-\" for stdout)")
+        .action((x, c) => c.copy(outputFile = if (!x.getPath.isEmpty && x.getPath != "-") Some(x) else None))
+
+      opt[Unit]('l', "list")
+        .text("List tests instead of building them")
+        .action((_, c) => c.copy(list = true))
+
+      opt[Unit]('v', "verbose")
+        .text("Print verbose information to stderr")
+        .action((_, c) => c.copy(verbose = true))
+
+      opt[Seq[String]]('f', "filter")
+        .text("Only consider tests which match at least one filter regex")
+        .unbounded()
+        .action((x, c) => c.copy(filters = c.filters ++ x.map(_.r)))
+
+      opt[Seq[String]]('x', "exclude")
+        .text("Ignore tests which match at least one exclusion regex")
+        .unbounded()
+        .action((x, c) => c.copy(excludes = c.excludes ++ x.map(_.r)))
+
+      // Do not `sys.exit` on `help` to facilitate testing.
+      override def terminate(exitState: Either[String, Unit]): Unit = {
+        shouldExit = true
+      }
+    }
+
+    // Parse the command line options.
+    val config = parser.parse(args, Config()) match {
+      case Some(config) => config
+      case None         => return
+    }
+    if (shouldExit)
+      return
+
+    // Define the handler that will be called for each discovered unit test, and
+    // which will decide whether the test is generated or not.
+    def handler(className: String, gen: () => Unit): Unit = {
+      // If none of the inclusion filters match, skip this test.
+      if (!config.filters.isEmpty && !config.filters.exists(_.findFirstMatchIn(className).isDefined)) {
+        Console.err.println(f"Skipping ${className} (does not match filter)")
+        return
+      }
+
+      // If any of the exclusion filters match, skip this test.
+      if (config.excludes.exists(_.findFirstMatchIn(className).isDefined)) {
+        Console.err.println(f"Skipping ${className} (matches exclude filter)")
+        return
+      }
+
+      // If we are just listing tests, print the class name and skip this test.
+      if (config.list) {
+        println(className)
+        return
+      }
+
+      // Otherwise generate the test.
+      if (config.verbose)
+        Console.err.println(f"Building ${className}")
+      gen()
+    }
+
+    // Generate the unit tests.
+    class AllUnitTests extends RawModule {
+      DiscoverUnitTests(handler)
+    }
+    val chirrtl = ChiselStage.emitCHIRRTL(new AllUnitTests)
+    if (config.list)
+      return
+
+    // Write the result to the output.
+    config.outputFile match {
+      case Some(file) =>
+        val w = new PrintWriter(file)
+        try {
+          w.write(chirrtl)
+        } finally {
+          w.close()
+        }
+      case None => print(chirrtl)
+    }
+  }
```

Comment:
This is fine. The PR could unify this handling by defining a single thing to write to and then always using that. Also, use of `PrintStream` is likely preferred. That also has the benefit of `System.Out` being an instance of `PrintStream` already.

Reply:
:+1: done